File: backend/app/models/hotspot_annotations.py
"""Contracts for human review and knowledge-base integration."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class AnnotationModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class AnnotationUpdateRequest(AnnotationModel):
    review_status: Literal["reviewed", "watch", "ignored"] = "reviewed"
    topic_category: str | None = Field(default=None, max_length=80)
    tags: list[str] = Field(default_factory=list, max_length=20)
    hospitality_relevance: Literal[
        "relevant", "possibly_relevant", "irrelevant"
    ] | None = None
    decision_lane: Literal[
        "opportunity",
        "context_signal",
        "risk_watch",
        "reference_material",
        "background",
    ] | None = None
    notes: str | None = Field(default=None, max_length=4000)
    summary_override: str | None = Field(default=None, max_length=2000)

    @field_validator("tags")
    @classmethod
    def normalize_tags(cls, values: list[str]) -> list[str]:
        normalized: list[str] = []
        for value in values:
            tag = value.strip()
            if not tag:
                continue
            if len(tag) > 32:
                raise ValueError("单个标签不能超过 32 个字符")
            if tag not in normalized:
                normalized.append(tag)
        return normalized

    @field_validator("topic_category", "notes", "summary_override")
    @classmethod
    def normalize_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        return stripped or None


class EventAnnotationView(AnnotationModel):
    annotation_id: str
    trend_series_id: str
    source_event_id: str
    revision: int
    review_status: str
    topic_category: str | None = None
    tags: list[str]
    hospitality_relevance: str | None = None
    decision_lane: str | None = None
    notes: str | None = None
    summary_override: str | None = None
    updated_by: str
    created_at: str


class KnowledgeSyncView(AnnotationModel):
    sync_id: str
    annotation_id: str
    trend_series_id: str
    source_event_id: str
    provider: str
    knowledge_base_id: str
    knowledge_id: str | None = None
    content_hash: str
    status: str
    operation: str
    parse_status: str | None = None
    error_kind: str | None = None
    error_message: str | None = None
    started_at: str
    finished_at: str


class KnowledgeSyncResponse(AnnotationModel):
    sync: KnowledgeSyncView
    skipped_existing: bool = False


class KnowledgeSearchRequest(AnnotationModel):
    query: str | None = Field(default=None, max_length=300)
    match_count: int = Field(default=5, ge=1, le=20)

    @field_validator("query")
    @classmethod
    def normalize_query(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        return stripped or None


class KnowledgeSearchHit(AnnotationModel):
    chunk_id: str
    content: str
    knowledge_id: str
    knowledge_title: str | None = None
    score: float | None = None
    chunk_index: int | None = None
    source: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class KnowledgeSearchResponse(AnnotationModel):
    provider: str
    query: str
    knowledge_base_id: str
    hits: list[KnowledgeSearchHit]


class KnowledgeIntegrationStatus(AnnotationModel):
    provider: Literal["weknora"] = "weknora"
    configured: bool
    base_url_configured: bool
    api_key_configured: bool
    knowledge_base_id_configured: bool

File: frontend/src/utils/hotspotAnnotations.js
const DEFAULT_ANNOTATION = {
  review_status: 'reviewed',
  topic_category: '',
  tags: '',
  hospitality_relevance: '',
  decision_lane: '',
  notes: '',
  summary_override: '',
}

export function annotationDraft(event) {
  const annotation = event?.annotation
  const classification = event?.classification
  return {
    ...DEFAULT_ANNOTATION,
    topic_category: annotation?.topic_category || classification?.topic_category || '',
    tags: (annotation?.tags || classification?.tags || []).join('，'),
    hospitality_relevance:
      annotation?.hospitality_relevance || classification?.hospitality_relevance || '',
    review_status: annotation?.review_status || DEFAULT_ANNOTATION.review_status,
    decision_lane: annotation?.decision_lane || '',
    notes: annotation?.notes || '',
    summary_override: annotation?.summary_override || '',
  }
}

export function annotationPayload(draft) {
  const tags = String(draft.tags || '')
    .split(/[,，]/)
    .map(tag => tag.trim())
    .filter((tag, index, values) => tag && values.indexOf(tag) === index)

  return {
    review_status: draft.review_status,
    topic_category: draft.topic_category?.trim() || null,
    tags,
    hospitality_relevance: draft.hospitality_relevance || null,
    decision_lane: draft.decision_lane || null,
    notes: draft.notes?.trim() || null,
    summary_override: draft.summary_override?.trim() || null,
  }
}

export function knowledgeConfigurationText(status) {
  if (status?.configured) return 'WeKnora 已连接'
  const missing = []
  if (!status?.base_url_configured) missing.push('服务地址')
  if (!status?.api_key_configured) missing.push('API Key')
  if (!status?.knowledge_base_id_configured) missing.push('知识库 ID')
  return `待配置：${missing.join('、') || '连接参数'}`
}
